src/lib.mo

```diff
@@ -0,0 +1,327 @@
+      ignore addPullValue("rts_mutator_instructions", func() = Prim.rts_mutator_instructions());
+      ignore addPullValue("rts_collector_instructions", func() = Prim.rts_collector_instructions());
+      ignore addPullValue("stablememory_size", func() = Nat64.toNat(StableMemory.size()));
+    };
+
+    func removeValue(id : Nat) : () = values.put(id, null);
+
+    /// Dump all current stats to array
+    public func dump() : [(Text, Nat)] {
+      let result = Vector.Vector<(Text, Nat)>();
+      for (v in values.vals()) {
+        switch (v) {
+          case (?value) Vector.addFromIter(result, Iter.fromArray(value.dump()));
+          case (null) {};
+        };
+      };
+      Vector.toArray(result);
+    };
+
+    func renderSingle(name : Text, value : Text, timestamp : Text) : Text = name # " " # value # " " # timestamp # "\n";
+
+    /// Render all current stats to prometheus format
+    public func renderExposition() : Text {
+      let timestamp = Int.toText(now() / 1000000);
+      var res = "";
+      for ((name, value) in dump().vals()) {
+        res #= renderSingle(name, Nat.toText(value), timestamp);
+      };
+      res;
+    };
+
+    /// Dump all values, marked as stable, to stable data structure
+    public func share() : StableData {
+      var res : StableData = null;
+      for (value in values.vals()) {
+        switch (value) {
+          case (?v) switch (v.share()) {
+            case (?data) {
+              res := AssocList.replace(res, v.prefix, Text.equal, ?data).0;
+            };
+            case (_) {};
+          };
+          case (null) {};
+        };
+      };
+      res;
+    };
+
+    /// Patch all values with stable data
+    public func unshare(data : StableData) : () {
+      for (value in values.vals()) {
+        switch (value) {
+          case (?v) switch (AssocList.find(data, v.prefix, Text.equal)) {
+            case (?data) v.unshare(data);
+            case (_) {};
+          };
+          case (_) {};
+        };
+      };
+    };
+
+  };
+
+  class PullValue(prefix_ : Text, pull : () -> Nat) {
+    public let prefix = prefix_;
+
+    public func dump() : [(Text, Nat)] = [(prefix # "{}", pull())];
+
+    public func share() : ?StableDataItem = null;
+    public func unshare(data : StableDataItem) = ();
+  };
+
+  class CounterValue(prefix_ : Text, isStable : Bool) {
+    public let prefix = prefix_;
+
+    public var value = 0;
+
+    public func add(n : Nat) { value += n };
+    public func set(n : Nat) { value := n };
+
+    public func dump() : [(Text, Nat)] = [(prefix # "{}", value)];
+
+    public func share() : ?StableDataItem {
+      if (not isStable) return null;
+      ? #counter(value);
+    };
+    public func unshare(data : StableDataItem) = switch (data, isStable) {
+      case (#counter x, true) value := x;
+      case (_) {};
+    };
+  };
+
+  class GaugeValue(prefix_ : Text, buckets_ : ?[Nat], now : () -> Int, watermarkResetIntervalSeconds : Nat, isStable : Bool) {
```

Comment:
What is the reason of passing in `now()` as a function as opposed to calling `Time.now()` directly inside this class?

Reply:
If we pass in a function then I would probably use a function `() -> Nat64` and change the internal timestamp and reset interval to type `Nat64`. It reduces cost of the update function by 50%. The function passed in would be `Nat64.fromIntWrap(Time.now())`.

Reply:
It would also be good to find a way to call `Time.now()` only once per `update`, not twice for high and low watermark. Because Time.now costs ~1,000 instructions.

Reply:
For tests. when launcheing code with `wasmtime` it traps with error `cannot get system time when running locally`. So I made this simple trick to change this function from the outside

Reply:
Refactored how we work with current time. For gauge value  now() will be called only once, also all timetamps are now in Nat64